File: cpp/cpp_100-999/318.cc
// C++
// leetcode 318
// https://leetcode-cn.com/problems/maximum-product-of-word-lengths/

class Solution {
public:
    int full = (1 << 26) - 1;

    int maxProduct(vector<string>& words) {
        int res = 0;
        map<int , int > m;  // hashval , length  
        for(auto s : words) {
            int length = s.size();
            int hashForThisString = 0;
            for(auto c : s) {
                hashForThisString = hashForThisString | (1 << (c - 'a'));
                if(hashForThisString == full)
                    break;
            }
            if(hashForThisString == full || length == 0)
                continue;
            for(auto [k, v] : m) {
                if((k & hashForThisString) == 0) {
                    // 没重复，可以试试
                    res = max(res, length * v);
                }
            }
            m[hashForThisString] = max(m[hashForThisString], length);
        }
        return  res;
    }
};

File: Go/Go_100-999/318.go
// Go
// leetcode 318
// https://leetcode.cn/problems/maximum-product-of-word-lengths/

func maxProduct(words []string) (res int) {
	cnt := make(map[int]int)
	full := 1<<26 - 1
	for _, s := range words {
		hashForThisWord := 0
		for _, c := range s {
			hashForThisWord |= 1<< (c-'a')
		}
		if hashForThisWord == full || hashForThisWord == 0 {
			continue
		}
		cnt[hashForThisWord] = max(cnt[hashForThisWord], len(s))
		for hash, length := range cnt {
			if hash & hashForThisWord != 0 {
				continue
			}
			res = max(res, length * len(s))
		}
	}
	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}